readthedocs/config/config.py

```diff
@@ -167,6 +167,44 @@ def catch_validation_error(self, key):
                 source_position=self.source_position,
             )
 
+    def pop(self, name, container, default, raise_ex):
+        """
+        Search and pop a key inside a dict.
+
+        This will pop the keys recursively if the container is empty.
+
+        :param name: the key name in a dotted form (key.innerkey)
+        :param container: a dictionary that contains the key
+        :param default: default value to return if the key doesn't exists
+        :param raise_ex: if True, raises an exception when a key is not found
+        """
+        key = name[0]
+        validate_dict(container)
+        if key in container:
+            if len(name) > 1:
+                value = self.pop(name[1:], container[key], default, raise_ex)
+                if not container[key]:
+                    container.pop(key)
+            else:
+                value = container.pop(key)
+            return value
+        if raise_ex:
+            raise ValidationError(key, VALUE_NOT_FOUND)
+        return default
+
+    def pop_config(self, key, *args):
+        """
+        Search and pop a key (recursively) from `self.raw_config`.
+
+        :param key: the key name in a dotted form (key.innerkey)
+        :param args: Optionally, it can receive a default value
+                     after the key, if no value is passed,
+                     it raises an exception when the key is not found.
+        """
+        raise_ex = not bool(args)
```

Comment:
So... I don't like much this _hack_, but I didn't find a better way to handle this. We can't put something like `default=None`, because we may need to default some values to None sometimes too. And I didn't want to add another parameter like `default=None, raise_ex`.

Reply:
Why can't the method just be:

```python
def pop_config(self, key, default=None, raise_ex=True):
    return self.pop(key.split('.'), self.raw_config, default, raise_ex
```

This still works if normal usage is:

```python
self.pop_config('formats', [])
```

Reply:
I didn't want to have another param here, but looks better than my hacky solution p:

Reply:
`raise_ex` should be False here by default,  so when we call `self.pop_config('key', 'default')` we don't need to pass another value like `self.pop_config('key', 'default', raise_ex=False)`. But when we don't provide a default value, we need to called it like `self.pop_config('key', raise_ex=False)`.

I'm trying to replicate the _same behaviour_ from https://docs.python.org/3/library/stdtypes.html#dict.pop. Looks like the implementation is inside the c code.

Reply:
It seems like `raise_ex=False` would be the default value then. Does than not solve your use case?

Reply:
Already implemented that in the last commit!

Reply:
`raise_ex = not bool(args)` and the rest of this method still seem like a non-standard way of doing a standard thing to me:

```python
def pop_config(self, key, default=None, raise_ex=False):
    if default is None:
        raise_ex = True
    return self.pop(key.split('.'), self.raw_config, default, raise_ex)
```

This seems to be the same logic without having to manually handle `*args` and hacking `raise_ex`

Reply:
Actually, we have more cases where `default=None` and shouldn't raise an exception, we only have one use of `default=None` and `raise_ex=True`. I don't think is a good idea to modify args based on other args.

This is the current implementation

https://github.com/stsewd/readthedocs.org/blob/c17e81c975a0639d56fc6028e90f80c5598bd3e7/readthedocs/config/config.py#L196-L204